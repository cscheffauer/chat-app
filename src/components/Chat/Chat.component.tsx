import React, { FormEvent, KeyboardEvent, useState, useEffect, useRef } from 'react'

import Message from './Message.component'
import './Chat.scss';

import { MessageType } from '../../models/chat.model'

interface Props {
    sendMessage: (message: String) => void,
    editMessage: (messageid: String, text: String) => void,
    deleteMessage: (messageid: String) => void,
    messages: Array<MessageType>,
    userid: String,
}

const Chat = ({ sendMessage, editMessage, messages, ...PropsFromParent }: Props) => {
    const [message, setmessage] = useState("");
    const [editmode, seteditmode] = useState(false);
    const [messageidtoedit, setmessageidtoedit] = useState("" as String);
    const [originalmessage, setoriginalmessage] = useState('' as String);

    const messageInput = useRef<HTMLInputElement>(null);        //ref for messageInput
    const scrollArea = useRef<HTMLDivElement>(null);        //ref for scrollArea

    useEffect(() => {
        if (messageInput.current !== null) {
            messageInput.current.focus();           //focus on the messageInput after messages changed
        }
        if (scrollArea.current !== null) {
            scrollArea.current.scrollTop = scrollArea.current.scrollHeight;     //to scroll to the bottom after messages changed
        }
    }, [messages]);


    const handleChange = (event: FormEvent<HTMLInputElement>) => {
        event.preventDefault();
        setmessage(event.currentTarget.value)
    }

    const handleSubmit = (event: FormEvent | KeyboardEvent) => {
        event.preventDefault();
        if (message.length > 0) {
            if (editmode) {
                if (originalmessage !== message) {
                    editMessage(messageidtoedit, message);
                    cancelEditMode();
                }
            } else {
                sendMessage(message);
                setmessage("");
            }
        }
    }

    const switchToEditMode = (messageid: String, messagetext: String) => {
        setmessageidtoedit(messageid);
        setoriginalmessage(messagetext);
        setmessage(messagetext as string);
        seteditmode(true);
        if (messageInput.current !== null) {
            messageInput.current.focus();           //focus on the messageInput to edit the message
        }
    }

    const cancelEditMode = () => {
        seteditmode(false);
        setmessage("");
    }

    return (
        <div className="chatWindow">
            <div ref={scrollArea} className="chatMessages">
                {
                    messages.map((message, i) => <Message key={i} message={message} switchToEditMode={switchToEditMode} {...PropsFromParent} />)
                }
            </div>
            <div className="chatInput">
                <form onSubmit={handleSubmit}>
                    <input ref={messageInput} placeholder={"Message"} onChange={handleChange} onBlur={cancelEditMode} value={message} type="text"></input>
                    {editmode && <span>âœŽ</span>}
                </form>
            </div>
        </div>

    )
}

export default Chat
