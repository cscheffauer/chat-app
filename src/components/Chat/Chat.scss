@import '../../utils/variables';

.chatWindow {
	height: 84vh; /* Fallback for browsers that do not support Custom Properties */
	height: calc(var(--vh, 1vh) * 84);
	background-color: $color-chat-window;
	display: flex;
	flex-direction: column;
	align-content: space-between;
	border-left: $border-style-tabs;
	border-right: $border-style-tabs;

	.chatMessages {
		flex: 1;
		overflow-y: auto;
		width: 97%;
		margin-top: 20px;
	}
}
