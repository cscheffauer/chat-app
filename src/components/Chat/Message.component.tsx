import React from 'react';
import Linkify from 'react-linkify';
import './Chat.scss';

interface Props {
    message: {
        sent: Date,
        username: String,
        text: String,
    }
}

const Message = ({ message }: Props) => {
    const sentDate = new Date(message.sent);
    const sent = sentDate.getHours() + ":" + sentDate.getMinutes();
    return (
        <div className="message">
            <p><span style={{ color: "#00002c", fontWeight: 500 }}>{message.username}</span><span style={{ color: "#9da2aa", marginLeft: 10 }}>{sent}</span> </p>
            <Linkify>
                <p style={{ lineHeight: '150%', color: message.username === "Meetingbot" ? 'grey' : 'black' }}>{message.text}</p>
            </Linkify>
        </div >
    )
}

export default Message
