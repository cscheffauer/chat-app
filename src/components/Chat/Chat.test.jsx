import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';

import Chat from './Chat.component';

// TODO: rework test with React Context https://stackoverflow.com/questions/54292298/jest-mock-react-context

describe('<Chat />', () => {
	let wrapper;
	const useridMock = '1234';
	const messagesMock = [
		{ username: 'Max', text: 'Hallo', sent: new Date('2020-01-01'), userid: '1', messageid: '1', state: '' },
		{ username: 'John', text: 'Hi', sent: new Date('2020-01-01'), userid: '2', messageid: '2', state: 'DELETED' },
		{ username: 'Jeffrey', text: "What's up?", sent: new Date('2020-01-01'), userid: '3', messageid: '3', state: 'EDITED' },
	];

	const sendMessageMock = jest.fn();
	const editMessage = jest.fn();
	const deleteMessage = jest.fn();

	beforeEach(() => {
		wrapper = mount(<Chat sendMessage={sendMessageMock} editMessage={editMessage} deleteMessage={deleteMessage} userid={useridMock} messages={messagesMock} />);
	});

	afterEach(() => {
		jest.clearAllMocks();
	});

	it('expect to render Chat component', () => {
		expect(wrapper).toMatchSnapshot();
	});
});
