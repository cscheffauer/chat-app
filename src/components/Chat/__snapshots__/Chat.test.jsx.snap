// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chat /> expect to render Chat component 1`] = `
<Chat
  deleteMessage={[MockFunction]}
  editMessage={[MockFunction]}
  messages={
    Array [
      Object {
        "messageid": "1",
        "sent": 2020-01-01T00:00:00.000Z,
        "state": "",
        "text": "Hallo",
        "userid": "1",
        "username": "Max",
      },
      Object {
        "messageid": "2",
        "sent": 2020-01-01T00:00:00.000Z,
        "state": "DELETED",
        "text": "Hi",
        "userid": "2",
        "username": "John",
      },
      Object {
        "messageid": "3",
        "sent": 2020-01-01T00:00:00.000Z,
        "state": "EDITED",
        "text": "What's up?",
        "userid": "3",
        "username": "Jeffrey",
      },
    ]
  }
  sendMessage={[MockFunction]}
  userid="1234"
>
  <div
    className="chatWindow"
  >
    <div
      className="chatMessages"
    >
      <Message
        key="1"
        message={
          Object {
            "messageid": "1",
            "sent": 2020-01-01T00:00:00.000Z,
            "state": "",
            "text": "Hallo",
            "userid": "1",
            "username": "Max",
          }
        }
        switchToEditMode={[Function]}
      >
        <div
          className="message"
        >
          <p>
            <span
              className="sender"
            >
              Max
            </span>
            <span
              className="info"
            >
              01:00
            </span>
          </p>
          <Linkify
            componentDecorator={[Function]}
            hrefDecorator={[Function]}
            matchDecorator={[Function]}
            textDecorator={[Function]}
          >
            <p
              key="0"
              style={
                Object {
                  "color": "black",
                  "lineHeight": "150%",
                }
              }
            >
              Hallo
            </p>
          </Linkify>
        </div>
      </Message>
      <Message
        key="2"
        message={
          Object {
            "messageid": "2",
            "sent": 2020-01-01T00:00:00.000Z,
            "state": "DELETED",
            "text": "Hi",
            "userid": "2",
            "username": "John",
          }
        }
        switchToEditMode={[Function]}
      >
        <div
          className="message"
        >
          <p
            className="deleted"
          >
            Message has been deleted by 
            John
          </p>
        </div>
      </Message>
      <Message
        key="3"
        message={
          Object {
            "messageid": "3",
            "sent": 2020-01-01T00:00:00.000Z,
            "state": "EDITED",
            "text": "What's up?",
            "userid": "3",
            "username": "Jeffrey",
          }
        }
        switchToEditMode={[Function]}
      >
        <div
          className="message"
        >
          <p>
            <span
              className="sender"
            >
              Jeffrey
            </span>
            <span
              className="info"
            >
              01:00
            </span>
            <span
              className="info"
            >
              | edited
            </span>
          </p>
          <Linkify
            componentDecorator={[Function]}
            hrefDecorator={[Function]}
            matchDecorator={[Function]}
            textDecorator={[Function]}
          >
            <p
              key="0"
              style={
                Object {
                  "color": "black",
                  "lineHeight": "150%",
                }
              }
            >
              What's up?
            </p>
          </Linkify>
        </div>
      </Message>
    </div>
    <ChatInput
      parentProps={
        Object {
          "cancelEditMode": [Function],
          "editmode": false,
          "message": "",
          "messageRef": Object {
            "current": <input
              placeholder="Message"
              type="text"
              value=""
            />,
          },
          "messageidtoedit": "",
          "originalmessage": "",
          "setmessage": [Function],
        }
      }
    >
      <div
        className="chatInput"
      >
        <form
          onSubmit={[Function]}
        >
          <ImageUpload
            openImagePreview={[Function]}
          >
            <label
              className="imageupload"
              htmlFor="imageupload"
            >
              ðŸ“¸
            </label>
            <input
              accept="image/png, image/jpeg"
              id="imageupload"
              multiple={false}
              onChange={[Function]}
              style={
                Object {
                  "display": "none",
                }
              }
              type="file"
            />
          </ImageUpload>
          <input
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Message"
            type="text"
            value=""
          />
        </form>
      </div>
    </ChatInput>
  </div>
</Chat>
`;
